logLikelihoodHierarchical.m

```diff
@@ -0,0 +1,271 @@
+if nargout > 1
+    gradlLH = zeros(n_theta,1);
+    if nargout >  2
+        if(strcmp(options.distribution,'laplace'))
+            error('No user supplied Fisher information matrix available for the Laplace distribution.');
+        else
+            HlLH = zeros(n_theta,n_theta);
+        end
+    end
+end
+
+% Initialization for for scaling and noise parameters
+s = zeros(1,n_y,n_r,n_e); % vector including scaling factors
+noise = zeros(1,n_y,n_r,n_e); % vector including noises
+
+if nargout > 1
+    ds = zeros(1,n_y,n_theta,n_r,n_e);
+end
+
+% if ~checkValidityOptions(D,options)
+%     error(['options not valid, all oberservables/experiments that share a' ...
+%         'scaling parameter need to also share the noise parameter'])
+% end
+
+%% OPTIMAL VALUES FOR THE scaling FACTORS
+for ie = 1:numel(options.expgroups_scaling)
+    ind_e = options.expgroups_scaling{ie};
+    n_r = size(D(ind_e(1)).my,3);
+    for iy = 1:numel(options.obsgroups_scaling)
+        scale = options.scale(options.obsgroups_scaling{iy}(1));
+        ind_y = options.obsgroups_scaling{iy};
+        
+        switch options.distribution
+            case {'normal'}
+                temps = optimalScaling(ind_y,simulation(ind_e),D(ind_e),options,scale);
+                for ir = 1:n_r
+                    s(:,ind_y,ir,ind_e) = temps(:,:,ir);
+                end
+            case {'laplace'}
+                if nargout > 1
+                    [temps,tempds] = ...
+                        optimalScaling(ind_y,simulation(ind_e),D(ind_e),options,scale);
+                    for ir = 1:n_r
+                        s(:,ind_y,ir,ind_e) = temps(:,:,ir);
+                        for iiy = 1:numel(ind_y)
+                            for iie = 1:numel(ind_e)
+                                ds(:,ind_y(iiy),:,ir,ind_e(iie)) = tempds(:,:,:,ir);
+                            end
+                        end
+                    end
+                else
+                    temps = optimalScaling(ind_y,simulation(ind_e),D(ind_e),options,scale);
+                    for ir = 1:n_r
+                        s(:,ind_y,ir,ind_e) = temps(:,:,ir);
+                    end
+                end
+        end
+    end
+end
+
+%% OPTIMAL VALUES FOR THE noiseS
+for ie = 1:numel(options.expgroups_noise)
+    ind_e = options.expgroups_noise{ie};
+    n_r = size(D(ind_e(1)).my,3);
+    for iy = 1:numel(options.obsgroups_noise)
+        scale = options.scale((options.obsgroups_noise{iy}(1)));
+        ind_y = options.obsgroups_noise{iy};
+        ind_e = options.expgroups_noise{ie};
+        [tempnoise] = ...
+            optimalNoise(ind_y,simulation(ind_e),D(ind_e),options,s(:,:,:,ind_e),scale);
+        for ir = 1:n_r
+            noise(:,ind_y,ir,ind_e)=tempnoise(:,:,ir);
+        end
+        
+    end
+end
+
+%% save scaling and noise parameters
+if options.save
+    save('analytical_results.mat','s','noise');
+end
+
+%% COMPUTING lLH, gradlLH, HessianlLH
+switch options.distribution
```

Comment:
as for logLikelihood_BachmannJakStat, there is a large overlap in the code for laplace/gaussian